Align isolated conformances with the current proposal

DougGregor opened the pull request:
Make several changes to bring the isolated conformances implementation into alignment with the latest proposal:
* Use global-actor syntax to specify isolation conformances, e.g., `@MainActor P`, rather than `isolated P`
* Allow conformance isolation to differ from that of the conforming type
* Infer `@MainActor` on conformances when SE-0466 is enabled, which can be suppressed by `nonisolated`
* Test printing of conformance isolation

DougGregor committed:
[Isolated conformances] Change syntax to @<global actor type> P

Instead of using the `isolated P` syntax, switch to specifying the
global actor type directly, e.g.,

   class MyClass: @MainActor MyProto { ... }

No functionality change at this point

DougGregor committed:
Print isolated conformances in textual interfaces

DougGregor committed:
Suggest global actor isolation rather than 'isolated' for conformances

Update diagnostic and Fix-It to suggest global actor isolation on a
conformance (e.g, `@MainActor P`) rather than `isolated`.

DougGregor committed:
[Isolated conformances] Allow conformance isolation to differ from the type's

With the move to explicitly specifying the global actor for an isolated
conformance, we can now have conformances whose isolation differs from
that of the type, including having actors with global-actor-isolated
conformances. Introduce this generalization to match the proposal, and
update/add tests accordingly.

DougGregor committed:
[Isolated conformances] Start parsing 'nonisolated' on conformances

It doesn't mean anything yet, but parse it and make sure it reproduces
in the module interface.

DougGregor committed:
Simplify APIs for accessing conformance isolation

The NormalProtocolConformance APIs for checking for an explicitly-written
isolation on a conformance were easy to get to, and the real semantic
API was buried in the type checker, leading to some unprincipled
checking. Instead, create a central ProtocolConformance::getIsolation()
to get the (semantic) actor isolation, and let that be the only place
that will access the explicitly-written global actor isolation for a
conformance. Update all call sites appropriately.

DougGregor committed:
[Isolated conformances] Infer main actor isolation under UnspecifiedMeansMainActorIsolated

When code in the current module defaults to main actor (under SE-0466),
also infer main-actor isolation for protocol conformances of main-actor
isolated types.

DougGregor commented:
https://github.com/swiftlang/swift-syntax/pull/3015

@swift-ci please smoke test

DougGregor commented:
https://github.com/swiftlang/swift-syntax/pull/3015

@swift-ci please build toolchain

DougGregor committed:
Don't infer isolated conformances when the feature isn't enabled

DougGregor commented:
https://github.com/swiftlang/swift-syntax/pull/3015

@swift-ci please smoke test

DougGregor commented:
https://github.com/swiftlang/swift-syntax/pull/3015

@swift-ci please build toolchain Windows

DougGregor commented:
@swift-ci please smoke test Windows

slavapestov commented on include/swift/AST/ConformanceAttributes.h:
```diff
@@ -28,9 +30,15 @@ struct ConformanceAttributes {
+  SourceLoc globalActorAtLoc;
+
+  /// The global actor type to which this conformance is isolated.
+  TypeExpr *globalActorType = nullptr;
```
Maybe this ought to be a TypeLoc?

DougGregor replied:
It could. We get a TypeExpr out of the CustomAttr itself, so we can either stick that into a TypeLoc or keep it as-is. This way it's only a single pointer in ConformanceAttributes vs. the two-pointer TypeLoc, so I didn't do it.

slavapestov commented on lib/AST/TypeCheckRequests.cpp:
```diff
@@ -1368,6 +1368,38 @@ void AssociatedConformanceRequest::cacheResult(
+bool ConformanceIsolationRequest::isCached() const {
+  // We only want to cache for global-actor-isolated conformances. For
+  // everything else, which is nearly every conformance, this request quickly
+  // returns "nonisolated" so there is no point in caching it.
```
This is what split caching is for

DougGregor replied:
Ahh yes, so it is! Thanks for the reminder---I don't think I've used this feature before.

slavapestov commented on include/swift/AST/ProtocolConformance.h:
```diff
@@ -546,6 +555,9 @@ class NormalProtocolConformance : public RootProtocolConformance,
   DeclContext *Context;
 
+  /// The global actor isolation for this conformance, if there is one.
+  TypeExpr *globalActorIsolation = nullptr;
```
Is this null most of the time? You can use a single bit to record the presence of this attribute, while storing the actual value in the request evaluator cache only when present.

DougGregor replied:
It's NULL most of the time. We could have a put and put it an ASTContext-managed side table, since it needs to be there at the point when the NormalProtocolConformance is created.

DougGregor commented:
Nnnnoooooooooo!!!! I force-pushed the wrong branch. Now it is time for me to go cry

DougGregor committed:
Rework storage for conformance isolation

Switch over to split caching for the conformance isolation request,
which optimizes for the common case where the conformance is
nonisolated. Also put the explicit global actor TypeExpr* in an
ASTContext side table, so we don't take a pointer's worth of storage
in every conformance.

For that side table, introduce a new "ASTContext::GlobalCache" that's
there only for side tables, so we don't have to go add get/set
operations to ASTContext and recompile the world every time we want to
add a side table like this.

Thanks, Slava!

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Do not infer conformance isolation to an isolated protocol

DougGregor commented:
@swift-ci please smoke test
